Error During Search - Item has already been added.

<!-- 
  Please read contribution guideline first: https://github.com/dnnsoftware/Dnn.Platform/blob/development/CONTRIBUTING.md 
  Any potential security issues should be sent to security@dnnsoftware.com, rather than posted on GitHub
-->
## Description 

```
2018-08-01 07:46:40.981+00:00 [RD00155D480831][D:2][T:63][ERROR] DotNetNuke.Services.Search.Internals.LuceneControllerImpl - System.ArgumentException: Item has already been added. Key in dictionary: 'Evoq.StructuredContent.Visualizer.Components.BusinessController, Evoq.StructuredContent.Visualizer' Key being added: 'Evoq.StructuredContent.Visualizer.Components.BusinessController, Evoq.StructuredContent.Visualizer'
at System.Collections.Hashtable.Insert(Object key, Object nvalue, Boolean add)
at DotNetNuke.Services.Search.Controllers.ModuleResultController.GetModuleSearchController(ModuleInfo module)
at DotNetNuke.Services.Search.Controllers.ModuleResultController.HasViewPermission(SearchResult searchResult)
at DotNetNuke.Services.Search.Controllers.SearchControllerImpl.HasPermissionToViewDoc(Document document, SearchQuery searchQuery)
at DotNetNuke.Services.Search.Internals.SearchSecurityTrimmer.PrepareScoreDocs()
at DotNetNuke.Services.Search.Internals.LuceneControllerImpl.Search(LuceneSearchContext searchContext, Int32 retryTimes)
2018-08-01 07:46:40.997+00:00 [RD00155D480831][D:2][T:63][ERROR] DotNetNuke.Services.Search.Internals.LuceneControllerImpl - Search Index Folder Is Not Available: Item has already been added. Key in dictionary: 'Evoq.StructuredContent.Visualizer.Components.BusinessController, Evoq.StructuredContent.Visualizer' Key being added: 'Evoq.StructuredContent.Visualizer.Components.BusinessController, Evoq.StructuredContent.Visualizer', Retry 1 time(s).
```

--- a/DNN Platform/Library/Services/Search/Controllers/ModuleResultController.cs
+++ b/DNN Platform/Library/Services/Search/Controllers/ModuleResultController.cs
@@ -53,6 +53,7 @@ namespace DotNetNuke.Services.Search.Controllers
         private static readonly ILog Logger = LoggerSource.Instance.GetLogger(typeof(ModuleResultController));
 
         private static Hashtable _moduleSearchControllers = new Hashtable();
+        private static object _threadLock = new object();
 
         #region Abstract Class Implmentation
 
@@ -187,15 +188,19 @@ namespace DotNetNuke.Services.Search.Controllers
                 return null;
             }
 
-            if (_moduleSearchControllers.ContainsKey(module.DesktopModule.BusinessControllerClass))
+            if (!_moduleSearchControllers.ContainsKey(module.DesktopModule.BusinessControllerClass))
             {
-                return _moduleSearchControllers[module.DesktopModule.BusinessControllerClass] as IModuleSearchResultController;
+                lock (_threadLock)
+                {
+                    if (!_moduleSearchControllers.ContainsKey(module.DesktopModule.BusinessControllerClass))
+                    {
+                        var controller = Reflection.CreateObject(module.DesktopModule.BusinessControllerClass, module.DesktopModule.BusinessControllerClass) as IModuleSearchResultController;
+                        _moduleSearchControllers.Add(module.DesktopModule.BusinessControllerClass, controller);
+                    }
+                }
             }
 
-            var controller = Reflection.CreateObject(module.DesktopModule.BusinessControllerClass, module.DesktopModule.BusinessControllerClass) as IModuleSearchResultController;
-            _moduleSearchControllers.Add(module.DesktopModule.BusinessControllerClass, controller);
-
-            return controller;
+            return _moduleSearchControllers[module.DesktopModule.BusinessControllerClass] as IModuleSearchResultController;
         }
 
         private bool ModuleIsAvailable(TabInfo tab, ModuleInfo module)

